text/0000-signed-transaction-payload/0000-signed-transaction-payload.md

```diff
@@ -0,0 +1,825 @@
+                                        </td>
+                                    </tr>
+                                    <tr>
+                                        <td>Tag</td>
+                                        <td>ByteArray[16]</td>
+                                        <td>The tag which is used to index the <i>Signed Transaction</i>'s hash.</td>
+                                    </tr>
+                                </table>
+                            </details>
+                        </td>
+                    </tr>
+                </table>
+            </details>
+        </td>
+    </tr>
+    <tr>
+        <td>Unlock Blocks Count</td>
+        <td>varint</td>
+        <td>The count of unlock blocks proceeding. Must match count of inputs specified.</td>
+    </tr>
+    <tr>
+        <td valign="top">Unlock Blocks <code>anyOf</code></td>
+        <td colspan="2">
+            <details open="true">
+                <summary>Signature Unlock Block</summary>
+                <blockquote>
+                Defines an unlock block containing signature(s) unlocking input(s).
+                </blockquote>
+                <table>
+                    <tr>
+                        <th>Name</th>
+                        <th>Type</th>
+                        <th>Description</th>
+                    </tr>
+                    <tr>
+                        <td>Unlock Type</td>
+                        <td>varint</td>
+                        <td>
+                            Set to <strong>value 0</strong> to denote a <i>Signature Unlock Block</i>.
+                        </td>
+                    </tr>
+                    <tr>
+                        <td valign="top">Signature <code>oneOf</code></td>
+                        <td colspan="2">
+                            <details>
+                                <summary>WOTS Signature</summary>
+                                <table>
+                                    <tr>
+                                        <th>Name</th>
+                                        <th>Type</th>
+                                        <th>Description</th>
+                                    </tr>
+                                    <tr>
+                                        <td>Signature Type</td>
+                                        <td>varint</td>
+                                        <td>
+                                            Set to <strong>value 0</strong> to denote a <i>WOTS Signature</i>.
+                                        </td>
+                                    </tr>
+                                    <tr>
+                                        <td>Signature</td>
+                                        <td>ByteArray</td>
+                                        <td>The signature signing the serialized <i>Unsigned Transaction</i>.</td>
```

Comment:
How to sign this with wots signature scheme using Kerl? AFAIK input message of kerl still have to be represented in trits.

Reply:
The legacy W-OTS just expects a 243-trit hash to be signed. One could use plain Keccak-384 for this and then convert the 48-byte hash to trits (exactly as it is currently done as part of Kerl). Then the current version of the ISS can be used exactly.

There will be a new RFC in the next days that describes this in detail, but as hinted above it is really not that complicated.

Reply:
Could I just convert the message to trits and use the current implementation of WOTS? If we use Keccak-384 that means another signing scheme implementation. It's not complicated for sure, but that doesn't mean it's not redundant.

Reply:
No you cannot, you need exactly 243 trits. If you convert the Message using `b1t6` you get an arbitrary amount of trits.
The method I've described above is exactly Kerl, just the first step that converts the input is skipped. That's also exactly why I've mentioned Keccak-384 (which is used for Kerl) and not some other hash function, to avoid the introduction of some new hashing function.

Reply:
Maybe I'm wrong here, but - When I sign a WOTS-signature - the only thing I actually use, is 1-3 (sec-level) balanced arrays of ints from -13 to 13. - aka "The'normalized bundle-hash" .. When creating a new signature-algo-rithm - it should be enough to 'recreate' this setup using any kind of algorithm - and o avoid th M-bug - one should avoid having the first valuee of any of the 3 set to 13. There should not be a need for exactly 243 trits at all- The trits used during signing, is from the private key, What is neede to sign is therefore only a balanced array so that the sum of each of the three sub-arrays is 0.

Reply:
> What I'm saying, is that the 'input-message' to be signed is an int-array of 81 ints ranging from -13 to 13.
Exactly, that's what I meant above! You need 81 trytes (-13...13) as an input. 81 trytes are exactly the 243 trits I've mentioned.

It is debatable whether the normalization is part of the signature scheme or part of the generation of the "input-message" to be signed. I'd usually consider this as part of the signature scheme as this is very special to the IOTA W-OTS and not used in other Winternitz signature schemes, but this is just a definition.

Reply:
The normalisation-step used today, is only a security-step to avoid exposing more than 50% of the key.
Signing avd validation fo the signature based on a pure bundle-hash would work just fine, but would be less secure, since you could easily end up exposing more than 80% of the key if the bundle-has was 'bad' for signing-purpose.
In that regard - however you chosse to calcualte those three parts to be signed - you should keep the normalisation as a security-spep, but you don't have to do it the same way when creating a new signature-schema. For instance - you could choose to have the new 'input' as a 24-based value (0-23) and normalize on 12 (or 11), and hash 4 more times to verify the public key (address) and that would only occure on the nodes when validating the signature.  24 is a more suidted number for binary operations - you would however make sure that the values ectracted to sign needs to be converted to 81 numbers from 0 to 23. This would be bit less secure that using all the 27 hashings for signing, but probably less expensive to calculate on the node.

Reply:
Actually - if you choose to use 24-based signing-schema - you could choose to start with 2 hashes before starting on the input - thus immideatly avoiding the M-bug, since you will never push the first signature-part as a clean copy of the private key :)

Reply:
> The normalisation-step used today, is only a security-step to avoid exposing more than 50% of the key.

Unfortunately, that is not correct. The normalization is a crucial and integral part to the security of the used W-OTS signature scheme. Without it, after the (first) signature has been exposed, an attacker has the signature for every bundle hash that is element-wise ≥ than the signed bundle hash. Even with just one signature, it would be trivial for an attacker to produce valid signatures for other messages. With normalization this is not possible. Traditional W-OTS schemes use a checksum that decreases when any element of the input increases. This has exactly the same effect, but it makes the signatures even longer and is therefore not 100% compatible with the current scheme.

However, you are right: How exactly the normalization is performed can be changed while still retaining 100% compatibility.
For example: Currently, the normalization even favors `M`s at the beginning of the bundle hash and this could be switched.
Also, you are of course correct that you can always make the codomain of the function used for the bundle hash smaller and still get a valid input for the signing. The main drawback with this is, that you (potentially considerably) reduce the security of the signature scheme: By reducing the codomain, it will be easier to find collisions, maybe even, because of some strange interplay of hash and normalization, considerably easier than just the reduction factor....

Using base24 is an interesting idea :thinking: although I am not really sure how much it would actually speed up the verification process as the bundle hash computation is only a small part of it.
One even faster alternative would for example be to consider each set bit of the hash as a 1/-1 and each un-set bit as a 0. However, this would reduce the security of the (not normalized) bundle hash from 243-trits to 243 bits...

Reply:
We will soon be publishing an RFC on Github that explains the "binary W-OTS" which we are discussing in more detail and provide alternatives.
@frohal, your feedback and ideas on those would then of course be greatly appreciated.

Reply:
I hope the "Binary-WOTS" could entirely based on Keccak. I was trying to implement the scheme with first step skipped (which is we doesn't convert trits to binary in Kerl). And it turned out to be a lot of works. Both WOTS and Kerl in bee would have to redesign since they are only considered ternary originally.

The new scheme I hope could be is we just need to convert the ternary seed to binary and just use Keccak do the hashing. So only Wots layer would need to rewrite.

Reply:
> One could use plain Keccak-384 for this and then convert the 48-byte hash to trits (exactly as it is currently done as part of Kerl)
Just a clarification here: you won't be able to use `b1t6` on the resulting Keccak-384 squeeze as it would yield 288 trits, unusable for signing. `iota.go`'s -> `kerl` -> `KerlBytesToTrits` will be used instead; do we have it documented somewhere?

Reply:
@Wollac I might be folling this on the wrong places, but I am not sure I can find that RFC you said that would be soon published.  Care to share a link to it ?